nova_vm/src/ecmascript/builtins/control_abstraction_objects/async_generator_objects.rs

```diff
@@ -2,100 +2,412 @@
+    }
+}
+
+impl TryFrom<Value> for AsyncGenerator<'_> {
+    type Error = ();
+
+    fn try_from(value: Value) -> Result<Self, Self::Error> {
+        if let Value::AsyncGenerator(value) = value {
+            Ok(value)
+        } else {
+            Err(())
+        }
+    }
+}
+
+impl TryFrom<Object> for AsyncGenerator<'_> {
+    type Error = ();
+
+    fn try_from(value: Object) -> Result<Self, Self::Error> {
+        if let Object::AsyncGenerator(value) = value {
+            Ok(value)
+        } else {
+            Err(())
+        }
+    }
+}
+
+impl InternalSlots for AsyncGenerator<'_> {
+    const DEFAULT_PROTOTYPE: ProtoIntrinsics = ProtoIntrinsics::AsyncGeneratorFunction;
+
+    fn get_backing_object(self, agent: &Agent) -> Option<OrdinaryObject<'static>> {
+        agent[self].object_index
+    }
+
+    fn set_backing_object(self, agent: &mut Agent, backing_object: OrdinaryObject<'static>) {
+        assert!(agent[self]
+            .object_index
+            .replace(backing_object.unbind())
+            .is_none());
+    }
+}
+
+impl InternalMethods for AsyncGenerator<'_> {}
+
+impl CreateHeapData<AsyncGeneratorHeapData, AsyncGenerator<'static>> for Heap {
+    fn create(&mut self, data: AsyncGeneratorHeapData) -> AsyncGenerator<'static> {
+        self.async_generators.push(Some(data));
+        AsyncGenerator(AsyncGeneratorIndex::last(&self.async_generators))
+    }
+}
+
+impl Index<AsyncGenerator<'_>> for Agent {
+    type Output = AsyncGeneratorHeapData;
+
+    fn index(&self, index: AsyncGenerator) -> &Self::Output {
+        &self.heap.async_generators[index]
+    }
+}
+
+impl IndexMut<AsyncGenerator<'_>> for Agent {
+    fn index_mut(&mut self, index: AsyncGenerator) -> &mut Self::Output {
+        &mut self.heap.async_generators[index]
+    }
+}
+
+impl Index<AsyncGenerator<'_>> for Vec<Option<AsyncGeneratorHeapData>> {
+    type Output = AsyncGeneratorHeapData;
+
+    fn index(&self, index: AsyncGenerator) -> &Self::Output {
+        self.get(index.get_index())
+            .expect("AsyncGenerator out of bounds")
+            .as_ref()
+            .expect("AsyncGenerator slot empty")
+    }
+}
+
+impl IndexMut<AsyncGenerator<'_>> for Vec<Option<AsyncGeneratorHeapData>> {
+    fn index_mut(&mut self, index: AsyncGenerator) -> &mut Self::Output {
+        self.get_mut(index.get_index())
+            .expect("AsyncGenerator out of bounds")
+            .as_mut()
+            .expect("AsyncGenerator slot empty")
+    }
+}
+
+#[derive(Debug, Default)]
+pub struct AsyncGeneratorHeapData {
+    pub(crate) object_index: Option<OrdinaryObject<'static>>,
+    pub(crate) async_generator_state: Option<AsyncGeneratorState>,
+}
+
+#[derive(Debug)]
+pub(crate) enum AsyncGeneratorState {
+    // SUSPENDED-START has `vm_or_args` set to Arguments, SUSPENDED-YIELD has it set to Vm.
+    Suspended {
+        state: SuspendedGeneratorState,
+        queue: VecDeque<AsyncGeneratorRequest>,
+    },
```

Comment:
In the `SUSPENDED-START` state, the queue will always be empty, except inside engine code between `%AsyncGeneratorPrototype%.next` steps 8-9.a. Would it make sense to split these states?

Reply:
Hmm, yeah might make sense.